src/engine/sync/abletonlink.h

```diff
@@ -0,0 +1,137 @@
+        return nullptr;
+    }
+
+    /// Notify a Syncable that their mode has changed. The Syncable must record
+    /// this mode and return the latest mode in response to getMode().
+    void setSyncMode(SyncMode mode) override;
+
+    /// Notify a Syncable that it is now the only currently-playing syncable.
+    void notifyUniquePlaying() override;
+
+    /// Notify a Syncable that they should sync phase.
+    void requestSync() override;
+
+    /// Must NEVER return a mode that was not set directly via
+    /// notifySyncModeChanged.
+    SyncMode getSyncMode() const override;
+
+    /// Only relevant for player Syncables.
+    bool isPlaying() const override;
+    bool isAudible() const override;
+    bool isQuantized() const override;
+
+    /// Gets the current speed of the syncable in bpm (bpm * rate slider), doesn't
+    /// include scratch or FF/REW values.
+    mixxx::Bpm getBpm() const override;
+
+    /// Gets the beat distance as a fraction from 0 to 1
+    double getBeatDistance() const override;
+
+    /// Gets the speed of the syncable if it was playing at 1.0 rate.
+    mixxx::Bpm getBaseBpm() const override;
+
+    /// The following functions are used to tell syncables about the state of the
+    /// current Sync Master.
+    /// Must never result in a call to
+    /// SyncableListener::notifyBeatDistanceChanged or signal loops could occur.
+    void updateLeaderBeatDistance(double beatDistance) override;
+
+    /// Enforces the immediate change of the beat distance of all Link peers
+    void forceUpdateLeaderBeatDistance(double beatDistance);
+
+    /// Must never result in a call to SyncableListener::notifyBpmChanged or
+    /// signal loops could occur.
+    void updateLeaderBpm(mixxx::Bpm bpm) override;
+
+    void notifyLeaderParamSource() override;
+
+    /// Combines the above three calls into one, since they are often set
+    /// simultaneously.  Avoids redundant recalculation that would occur by
+    /// using the three calls separately.
+    void reinitLeaderParams(double beatDistance, mixxx::Bpm baseBpm, mixxx::Bpm bpm) override;
+
+    /// Must never result in a call to
+    /// SyncableListener::notifyInstantaneousBpmChanged or signal loops could
+    /// occur.
+    void updateInstantaneousBpm(mixxx::Bpm bpm) override;
+
+    void onCallbackStart(int sampleRate,
+            int bufferSize,
+            std::chrono::microseconds absTimeWhenPrevOutputBufferReachesDac);
+    void onCallbackEnd(int sampleRate, int bufferSize);
+
+  private:
+    ableton::link::HostTimeFilter<MixxxClockRef> m_hostTimeFilter;
+    const QString m_group;
+    EngineSync* m_pEngineSync; // unowned, must outlive this.
+    SyncMode m_syncMode;
+
+    mixxx::Bpm m_oldTempo;
+
+    std::chrono::microseconds m_audioBufferTimeMicros;
+    std::chrono::microseconds m_absTimeWhenPrevOutputBufferReachesDac;
+    std::chrono::microseconds m_sampleTimeAtStartCallback;
+    std::chrono::microseconds m_timeAtStartCallback;
```

Comment:
can be simplified
* `m_audioBufferTimeMicros` is only set, never read
* `m_sampleTimeAtStartCallback` is only initialized, never changed nor read
* `m_timeAtStartCallback` is only set, never read
```suggestion
    std::chrono::microseconds m_absTimeWhenPrevOutputBufferReachesDac;
```